docs/ingress.md

```diff
@@ -0,0 +1,77 @@
+
+Most Ingress Controllers heavily rely on annotations to configure the ingress gateway. With Tyk Operator, our Ingress 
+ Controller prefers to reference a strongly typed custom resource template.
+
+## Motivation
+
+The standard Ingress resource is very basic and does not natively support many advanced capabilities that are required
+ for real-life use-cases. Despite this, the community have built tooling, capabilities, dependencies, and technical debt 
+ on top of the ingress resource. These all rely on abusing the metadata annotations, and this practice is apparent 
+ with the standard [Kubernetes NginX Ingress resource](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#annotations).
+
+In order to decouple, gain all the benefits of Kubernetes, offer a Native & consistent API, we introduced the Tyk
+ ApiDefinition custom resource, which brings Full-Lifecycle API Management capabilities to Kubernetes. A side effect of
+ this meant a potential trade-off between offering a K8s native experience & integrating with Ingress which would 
+ facilitate clean integration with 3rd party tooling built on-top of and dependent on the Ingress Resource.
+
+As a compromise & attempt to propose an alternative & more scalable solution, we have introduced the concept of the 
+ ingress template ApiDefinition resource. The Template ApiDefinition resource offers a means to extend the capabilities 
+ of the standard Ingress Resource, by merging features of the ingress specification with that of the template.
+
+## Sample Template ApiDefinition resource
+
+```yaml
+apiVersion: tyk.tyk.io/v1alpha1
+kind: ApiDefinition
+metadata:
+ name: myapideftemplate
+ labels:
+  template: "true"
+spec:
+ name: foo
+ protocol: http
+ use_keyless: true
+ proxy:
+  target_url: http://example.com
+```
+
+Pay particular attention to the ApiDefinition metadata. This specifies that we have an ApiDefinition object with the
+ label `ingress: true`.
+
+When applying this manifest, the ApiDefinition reconciler's predicate filter will skip reconciliation. This will allow
+ the ApiDefinition to be stored inside Kubernetes as a resource, but will not reconcile the ApiDefinition inside Tyk.
+
+All mandatory fields inside the ApiDefinition spec are still mandatory, but can be replaced with placeholders as they
+ will be overwritten by the Ingress reconciler.
+
+## Sample Ingress resource
+
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+ name: httpbin-ingress
+ annotations:
+  kubernetes.io/ingress.class: tyk # <----------------- REFERENCES TYK INGRESS CONTROLLER
+  tyk.io/template: myapideftemplate # <---------------- REFERENCE TO APIDEFINITION IN SAME NAMESPACE
+spec:
+ rules:
+  - http:
+     paths:
+      - path: /httpbin
+        pathType: Prefix
+        backend:
+         service:
+          name: httpbin
+          port:
+           number: 8000
```

Comment:
what does this do?  Is this a tyk api  with that name/port. Because that will be in conflict with the whole template concept, however we can leverage this to allow customers to provide a unique `service` name which will be the `APIDefinition` name, this will remove the need to use complex schemes to generate api names.

```diff
diff --git a/controllers/ingress_controller.go b/controllers/ingress_controller.go
index b14a274..70f79f6 100644
--- a/controllers/ingress_controller.go
+++ b/controllers/ingress_controller.go
@@ -148,7 +148,7 @@ func (r *IngressReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 				Spec: template.Spec,
 			}
 
-			api.Spec.Name = apiName
+			api.Spec.Name = p.Backend.ServiceName
 			api.Spec.Proxy.ListenPath = p.Path
 			api.Spec.Proxy.TargetURL = fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", p.Backend.ServiceName, namespacedName.Namespace, p.Backend.ServicePort.IntValue())
 
```

Reply:
The idea is that the template can be used/shared by many ingress resources and ingress rules.

Reply:
Isn't this what is happening? `Spec: template.Spec,`? we creating concrete resources here based on the template. And the name is replaced with actual name the user specified in the ingress backend.